src/ProjectSelectionManager.cpp

```diff
@@ -47,91 +47,51 @@ const ProjectSelectionManager &ProjectSelectionManager::Get(
    return Get( const_cast< AudacityProject & >( project ) );
 }
 
-ProjectSelectionManager::ProjectSelectionManager( AudacityProject &project )
-   : mProject{ project }
-{
-}
-
-ProjectSelectionManager::~ProjectSelectionManager() = default;
-
-bool ProjectSelectionManager::SnapSelection()
-{
-   auto &project = mProject;
-   auto &formats = ProjectNumericFormats::Get( project );
-   auto &settings = ProjectSettings::Get( project );
-   auto &window = ProjectWindow::Get( project );
-   auto snapTo = settings.GetSnapTo();
-   if (snapTo != SNAP_OFF) {
-      auto &viewInfo = ViewInfo::Get( project );
-      auto &selectedRegion = viewInfo.selectedRegion;
-      NumericConverter nc(NumericConverter::TIME,
-         formats.GetSelectionFormat(), 0, ProjectRate::Get(project).GetRate());
-      const bool nearest = (snapTo == SNAP_NEAREST);
-
-      const double oldt0 = selectedRegion.t0();
-      const double oldt1 = selectedRegion.t1();
-
-      nc.ValueToControls(oldt0, nearest);
-      nc.ControlsToValue();
-      const double t0 = nc.GetValue();
-
-      nc.ValueToControls(oldt1, nearest);
-      nc.ControlsToValue();
-      const double t1 = nc.GetValue();
-
-      if (t0 != oldt0 || t1 != oldt1) {
-         selectedRegion.setTimes(t0, t1);
-         return true;
-      }
-   }
-
-   return false;
-}
-
-double ProjectSelectionManager::AS_GetRate()
-{
-   return ProjectRate::Get(mProject).GetRate();
-}
+ProjectSelectionManager::ProjectSelectionManager(AudacityProject& project)
+    : mProject { project }
+    , mSnappingChangedSubscription { ProjectSnap::Get(project).Subscribe(
+         [this](auto&)
+         {
+            SnapSelection();
+         })}
+    , mTimeSignatureChangedSubscription {
+       ProjectTimeSignature::Get(project).Subscribe([this](auto&)
+                                                    { SnapSelection(); })
+    }
```

Comment:
Edge case.

On repeat review, I realized the subscriptions are incomplete.  Snapping may also be influenced by rate, and might need to change with project rate change, if you are so closely zoomed in, you can resolve samples.

Yes, unlikely.

But then I also thought:  there might be just one subscription here to ProjectSnap.  Project snap could listen to changes of rate and time signature (and whatever else might be snapped to in future) and translate it into a single message for change of parameters relevant to snap.

Reply:
Wait for the commit I will push in couple hours!

Reply:
Nah, I think it won't be really changed by the next commit.

Reply:
Right, the huge commit did not touch ProjectSelectionManager.

So the edge case could be fixed with one more subscription, to ProjectRate, there.